Remove ResourceList structure and usage.

glennw opened the pull request:
Previously, a resource list would be built at the start of the
frame. This would allow glyphs missing from the texture cache
to be rasterized.

However, sometimes (e.g. with subpixel antialiasing) it's not possible
to know right at the start of the frame exactly what needs to be
rasterized yet.

Instead, we now allow the resource cache to add requests during
the prepare_prim_for_render() stage. Primitives that request
resources are marked as needing a resolve operation. This allows
primitives that request resources to update their UV coords etc
before batching occurs (since texture IDs for resources must be
known before batching can occur).

This also lays most of the groundwork for running the resource
cache rasterizer as a separate thread in the future. This will be
used to ensure that if too many glyphs are requested in one frame,
WR can continue running without blocking, using older glyphs and
then re-render the scene when newer, high resolution glyphs are
available.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/523)
<!-- Reviewable:end -->

glennw commented:
r? @pcwalton (also @kvark particularly for the clip-related changes).

glennw commented:
Note: Let's get https://github.com/servo/webrender/pull/512 reviewed and merged before landing this (and I'll rebase mine on top of it). This can still be reviewed now though :)

pcwalton commented on webrender/src/prim_store.rs:
```diff
@@ -580,6 +566,74 @@ impl PrimitiveStore {
+                if let Some(address) = metadata.clip_index {
+                    let clip = self.gpu_data32.get_slice_mut(address, 6);
+                    let old = clip[5].data; //TODO: avoid retaining the screen rectangle
+                    clip[5] = GpuBlock32::from(ImageMaskInfo {
```
At some point I'd like to factor all these indices out into constants. (You don't have to do this now)

Maybe we could even make macros to make this stuff easier.

pcwalton commented on webrender/src/prim_store.rs:
```diff
@@ -580,6 +566,74 @@ impl PrimitiveStore {
+                                                               &text.glyph_indices, |index, uv0, uv1| {
+                        let dest_glyph = &mut dest_glyphs[index];
+                        let dest: &mut GlyphPrimitive = unsafe {
+                            mem::transmute(dest_glyph)
```
What is this transmute for?

glennw replied:
The existing data is a GpuBlock32 - I need to access it as the original type, in order to modify some fields of it. Is there a better way to express this (the From trait is used to create the GpuBlock32 initially)?

kvark commented on webrender/src/resource_cache.rs:
```diff
@@ -27,11 +26,24 @@ use webrender_traits::{GlyphDimensions, PipelineId, WebGLContextId};
 
+pub struct CacheItem {
+    pub texture_id: TextureId,
+    pub uv0: Point2D<f32>,
```
would be great to specify that this is in texels, not actual normalized UVs, either by better naming, typing, or just comments

glennw replied:
Added comment explaining that it is in texels, and why.

kvark commented on webrender/src/resource_cache.rs:
```diff
@@ -117,24 +129,43 @@ impl<K,V> ResourceClassCache<K,V> where K: Clone + Hash + Eq + Debug, V: Resourc
+}
+
+enum ResourceRequest {
+    Image(ImageResourceRequest),
```
I think it would be clearer to just have `Image(ImageKey, ImageRendeing)`, since `ImageResourceRequest` is not used in any other context anyway

glennw replied:
Fixed

kvark commented on webrender/src/resource_cache.rs:
```diff
@@ -117,24 +129,43 @@ impl<K,V> ResourceClassCache<K,V> where K: Clone + Hash + Eq + Debug, V: Resourc
 
     // TODO(pcwalton): Figure out the lifecycle of these.
-    webgl_textures: HashMap<WebGLContextId, TextureId, BuildHasherDefault<FnvHasher>>,
+    webgl_textures: HashMap<WebGLContextId, (TextureId, Size2D<i32>), BuildHasherDefault<FnvHasher>>,
```
This asks for a new type

glennw replied:
Fixed

kvark commented on webrender/src/resource_cache.rs:
```diff
@@ -212,79 +246,35 @@ impl ResourceCache {
+    pub fn request_image(&mut self,
+                         key: ImageKey,
+                         rendering: ImageRendering) {
+        debug_assert!(self.state == State::AddResources);
```
I assume this will be converted to a type system rule/restriction later on

kvark commented on webrender/src/resource_cache.rs:
```diff
@@ -347,10 +337,32 @@ impl ResourceCache {
+                let uv1 = Point2D::new(cache_item.pixel_rect.bottom_right.x as f32,
+                                       cache_item.pixel_rect.bottom_right.y as f32);
+                f(loop_index, uv0, uv1);
+                texture_id = cache_item.texture_id;
```
is this `texture_id` guaranteed to be the same for all glyphs?

glennw replied:
Yes.

kvark commented on webrender/src/tiling.rs:
```diff
@@ -456,12 +474,17 @@ impl AlphaBatcher {
 }
 
+struct RenderTargetContext<'a> {
+    layer_store: &'a Vec<StackingContext>,
```
Any reason not to use `&'a [StackingContext]`? I think that would reduce the indirections on access

glennw replied:
Fixed

kvark commented on webrender/src/tiling.rs:
```diff
@@ -2098,75 +2115,79 @@ impl FrameBuilder {
-                            color: color,
+            // Build list of passes, target allocs that each tile needs.
+            for screen_tile in screen_tiles {
+                let rect = screen_tile.rect;        // TODO(gw): Remove clone here
```
where is the clone?

glennw replied:
Removed comment

glennw commented:
@pcwalton @kvark  OK, addressed those review comments. I rebased kvark's PR on top of this locally, and the conflicts looked straightforward. So it probably doesn't matter which one lands first.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #519) made this pull request unmergeable. Please resolve the merge conflicts.

kvark commented:
In that case, you may add an assert to enforce this

> On Nov 3, 2016, at 18:33, Glenn Watson <notifications@github.com> wrote:
> 
> @glennw commented on this pull request.
> 
> In webrender/src/resource_cache.rs:
> 
> > +        let mut glyph_key = GlyphKey::new(font_key,
> +                                          size,
> +                                          blur_radius,
> +                                          0);
> +        let mut texture_id = TextureId::invalid();
> +        for (loop_index, glyph_index) in glyph_indices.iter().enumerate() {
> +            glyph_key.index = *glyph_index;
> +            let image_id = self.cached_glyphs.get(&glyph_key, self.current_frame_id);
> +            let cache_item = image_id.map(|image_id| self.texture_cache.get(image_id));
> +            if let Some(cache_item) = cache_item {
> +                let uv0 = Point2D::new(cache_item.pixel_rect.top_left.x as f32,
> +                                       cache_item.pixel_rect.top_left.y as f32);
> +                let uv1 = Point2D::new(cache_item.pixel_rect.bottom_right.x as f32,
> +                                       cache_item.pixel_rect.bottom_right.y as f32);
> +                f(loop_index, uv0, uv1);
> +                texture_id = cache_item.texture_id;
> Yes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 17a3629 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 17a36299ef371729c7dd2644f8586f39b2ae2e5c -->

bors-servo commented:
:hourglass: Testing commit 17a3629 with merge e7a66c5...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/webrender/builds/173136486)
